### ISSUE TITLE
JSON formatting issue

### ISSUE THREAD

**Issue by rounak610** [vllm-project/vllm#1191]
How to get the response from vLLM in a proper JSON format.
Does vLLM support outlines, guidance or jsonformer libraries?

**Comment by Kaffaljidhmah2** [comment]
The current version of vLLM supports adding `stop` or `stop_token_ids` so when LLMs generate those words/tokens, the generation is stopped, as in https://github.com/vllm-project/vllm/blob/main/vllm/sampling_params.py#L54.

I believe it's not hard to modify JsonFormer to support vLLM!

**Comment by viktor-ferenczi** [comment]
**Some best practices**
- Use a coding model if you can, like CodeLlama or WizardCoder, because they are better trained on JSON. Airoboros should also work reasonably well. If you need a smaller one try `microsoft/phi-1_5`.
- Request JSON output in the SYSTEM context or the prompt, for example: 
  * For base models end your prompt with : "The answer in JSON format is the following:"
  * For instruct or chat models: "Answer only in JSON. Do NOT write anything else."
- Models often like to respond with JSON in a code block (triple back-quotes). It is a good idea to find the first/last occurence of triple back-quotes and consider only the text in-between as the answer. Doing so would ignore any explanations may be added by the LLM. (They don't always follow prompts to the letter.)
- Validate the model's output (JSON formatting and contents) as much as possible and retry if not valid. 
- After multiple unsuccessful retries consider increasing the temperature slightly.
- Look for consistent mistakes the model makes in the formatting and implement automatic fixes for that. For example if it misses commas after items or the closing brace.
- Consider switching to more text-like or flatter output format, like YAML or TOML. They can also be parsed in Python and converted to JSON if you need the data in that format.

Contact me directly if you need more hands-on help.

**Comment by viktor-ferenczi** [comment]
I think the best may be to add support (API) for the integration of guidance libraries, then provide an adapter for each of the popular libraries separately.

@WoosukKwon What do you think?

**Comment by viktor-ferenczi** [comment]
Related to #535

**Comment by viktor-ferenczi** [comment]
Mostly the same as #288

**Comment by noamgat** [comment]
[LM Format Enforcer](https://github.com/noamgat/lm-format-enforcer) is a library that achieves this and supports vLLM.
There is already a [sample notebook](https://github.com/noamgat/lm-format-enforcer/blob/main/samples/colab_vllm_integration.ipynb) showing vLLM integration. It currently uses monkeypatching, that will be removed when the next vLLM version with the logits processing API will be released.

(Disclosure: I am the author of the library)

**Comment by arshadshk** [comment]
@noamgat does [LM Format Enforcer](https://github.com/noamgat/lm-format-enforcer) support api access to vllm ?

**Comment by noamgat** [comment]
> @noamgat does [LM Format Enforcer](https://github.com/noamgat/lm-format-enforcer) support api access to vllm ?

With LM Format Enforcer its the other way around - it integrates into the inference engine's pipeline. So if you have existing vLLM code, its easy to plug LMFE into it.

[vLLM example notebook](https://github.com/noamgat/lm-format-enforcer/blob/main/samples/colab_vllm_integration.ipynb)

**Comment by wdhitchc** [comment]
@noamgat 

since its not an integration with the VLLM Server though, I cant just have an LLM deployment use the format enforcer against that, I need to have the model loaded locally/ in the sample place as the format enforcer code....

**Comment by noamgat** [comment]
Yes, this is a known issue.
In order to solve it, we would need to be able to pass "logits processor instructions" in the network request of VLLM server.
I proposed something similar to huggingface-inference-server in [this draft PR meant for discussion](https://github.com/huggingface/text-generation-inference/pull/1274), but did not get a response from the team yet so I didn't proceed with it.

If vLLM would be interested in adopting a similar solution, it would cause the LMFE to also work with server / multi GPU deployments.

**Comment by viktor-ferenczi** [comment]
That would be nice. Support should be added for LMQL as well, I think.

**Comment by wdhitchc** [comment]
Is that not what's happening here @noamgat @viktor-ferenczi  https://github.com/vllm-project/vllm/pull/535

**Comment by noamgat** [comment]
A similar PR to #535 (disclosure: mine) was already merged. This is how LMFE works with vllm - [example notebook](https://github.com/noamgat/lm-format-enforcer/blob/main/samples/colab_vllm_integration.ipynb).

The challenge (that #535 does not cover) is how do we pass this information in a multiprocess / networked environment to the custom code. A serialization / "simple object" parametrization solution is required, which is what was proposed in the PR to huggingface-inference-server.

**Comment by wdhitchc** [comment]
@noamgat so maybe I'm not super clear on what's going on internally. I have not done my full due diligence and deep dived the code but....

I saw your PR that got in. I thought we just needed another one that utilizes what you put in, inside of the API/ openapi layer. My understanding was that the final comment on 535 was a request for him to rebase and utilize your code in the api layer. 


I was thinking about making a dumb solution for myself where I actually install LMfE into VLLM, then create a new api endpoint that is essentially a parameterized version of the notebook you shared.  I was thinking I'd have it inside the openapi compatible server and make it look like function calling. Not sure how efficient it would be but could get the job done quickly to accelerate me. This would probably just be a personal fork until the better solution gets in. I don't feel qualified to try and implement the correct pattern, but happy to give it a shot if you'll hold my hand a bit through the process.

**Comment by rlouf** [comment]
Outlines now provides a "fork" of vLLM's FastAPI deployment example, if that helps: https://outlines-dev.github.io/outlines/reference/vllm/

**Comment by noamgat** [comment]
> Outlines now provides a "fork" of vLLM's FastAPI deployment example, if that helps: https://outlines-dev.github.io/outlines/reference/vllm/

That is super cool! Congrats on the release!
If you want to avoid monkey patching the vLLM logits processor API, you can cache according to the generated token tuple instead of `seq_id`. I'm not sure how `seq_id` behaves with facilities such as beam searching etc.

**Comment by hmellor** [comment]
Support for guided decoding using `outlines` was merged in #2819
